File: src/python/rune.py
from entity import Entity
from caption import Caption

class _Rune(Entity):

    def __init__(self, engineRef, sprite):
        super(_Rune, self).__init__(engineRef, sprite, None)
        self.invincible = True
        self.name = self.sprite.name

        if self.name in engineRef.saveFlags:
            self.x = -100
            self.engineRef.destroyEntity(self)

    def apply(self):
        pass

    def update(self):
        if self.touches(self.engineRef.player):
            self.engineRef.destroyEntity(self)
            self.engineRef.addThing(Caption(self.engineRef, self.engineRef.font, '~1You got the %s Rune!' % self.element))
            self.engineRef.saveFlags[self.name] = 'True'
            self.apply()


class WaterRune(_Rune):
    element = property(lambda self: 'Water')


class FireRune(_Rune):
    element = property(lambda self: 'Fire')


class WindRune(_Rune):
    element = property(lambda self: 'Wind')


class CowardRune(_Rune):
    element = property(lambda self: 'Coward')
        
        
class BindingRune(_Rune):
    element = property(lambda self: 'Binding')


class StrengthRune(_Rune):

    def apply(self):
        self.engineRef.player.stats.att += 2

    element = property(lambda self: 'Strength')


class GuardRune(_Rune):

    def apply(self):
        self.engineRef.player.stats.pres += 2
        self.engineRef.player.stats.mres += 2

    element = property(lambda self: 'Guard')


class PowerRune(_Rune):

    def apply(self):
        self.engineRef.player.stats.mag += 2

    element = property(lambda self: 'Power')

File: src/winter/rune.ts
import { Caption } from "./caption.js"
import { Entity } from "./entity.js"
import { PyEngine, Sprite } from "./winter.js"

class Rune extends Entity {
    protected name: string

    constructor(
        engineRef: PyEngine,
        sprite: Sprite,
    ) {
        super(engineRef, sprite, {})
        this.invincible = true
        this.name = this.sprite.name

        if (engineRef.getSaveFlag(this.name)) {
            this.sprite.x = -100
            this.engineRef.destroyEntity(this)
        }
    }

    getElement() { return '' }
    apply() {}

    update() {
        if (this.touches(this.engineRef.getPlayerEntity())) {
            this.engineRef.destroyEntity(this)
            this.engineRef.addThing(new Caption(this.engineRef, this.engineRef.font, `~1You got the ${this.getElement()} Rune!`))
            this.engineRef.setSaveFlag(this.name, 'True')
            this.apply()
        }
    }
}

export class WaterRune extends Rune {
    getElement() { return 'Water' }
}

export class FireRune extends Rune {
    getElement() { return 'Fire' }
}

export class WindRune extends Rune {
    getElement() { return 'Wind' }
}

export class CowardRune extends Rune {
    getElement() { return 'Coward' }
}

export class BindingRune extends Rune {
    getElement() { return 'Binding' }
}

export class StrengthRune extends Rune {
    apply() {
        this.engineRef.getPlayerEntity().stats.att += 2
    }
    getElement() { return 'Strength' }
}

export class GuardRune extends Rune {
    apply() {
        this.engineRef.getPlayerEntity().stats.pres += 2
        this.engineRef.getPlayerEntity().stats.mres += 2
    }
    getElement() { return 'Guard' }
}

export class PowerRune extends Rune {
    apply() {
        this.engineRef.getPlayerEntity().stats.mag += 2
    }
    getElement() { return 'Power' }
}
